charts/kargo/templates/api/deployment.yaml

```diff
@@ -85,8 +85,11 @@ spec:
             - name: h2c
               containerPort: 8080
               protocol: TCP
+
 {{- if .Values.api.probes.enabled }}
```

Comment:
I think with the probes fully configurable, this option becomes redundant and should be removed from the template and the `values.yaml` file.

It is a breaking change, but it's a low-impact breaking change:

- Exists mainly to support local development.
- Anyone who may have disabled the probes already (in non-local dev envs), if this breaking change were to catch them by surprise, they'd be _gaining_ new probes rather than _losing_ anything. Whatever disruption that might cause should be easily identified and corrected.

I will of course mention a change such as this prominently in the release notes.

Reply:
It occurs to me that removing this option will require some slight modifications to `hack/tilt/values.dev.yaml`. You'll just have to nil out the probes for a couple of components.

Reply:
Alright, I'll just remove the tls-variable for the probe then

From what I see, the probes are already disabled in tilt?

* https://github.com/akuity/kargo/blob/main/hack/tilt/values.dev.yaml#L8
* https://github.com/akuity/kargo/blob/main/hack/tilt/values.dev.yaml#L22

Reply:
Just to mention it here: I'd personally leave the tls variable in the template. Removing it adds manual toil to deployments (like ours) because you'll be confused why the probe is failing when TLS is disabled.
in our case, we don't use TLS here because our ingress is already doing TLS

Reply:
> In tilt the probes are already disabled, so we shouldn't need to override the exec there

This suggests some misunderstanding.

> I think with the probes fully configurable, this option becomes redundant and should be removed from the template and the values.yaml file.

"This option," here referred to `api.probes.enabled`. There is no sense in having an `enabled` boolean when having or not having a liveness probe can simply depend on whether `api.probes.livenessProbe` is defined or not. Same can be said for the readiness probes.

Having said that, the following may change my line of thought...

> Just to mention it here: I'd personally leave the tls variable in the template. Removing it adds manual toil to deployments (like ours) because you'll be confused why the probe is failing when TLS is disabled.

That wasn't even on my radar when I initially reviewed, but you're right. I believe this exposes the fact that it's probably not so ideal to put the definitions of the probes _fully_ within the control of the user. It's not _just_ about TLS. It altogether eliminates any (current or future) possibility of probes being configured conditionally based on based on _other_ settings throughout the chart. TLS being enabled or not is just one such example.

This drives me toward thinking that we ought not let users define `livenessProbe` and `readinessProbe` in their entirety, but should, instead, surface individual options for things like `initialDelaySeconds`, `periodSeconds`, `timeoutSeconds`, etc., etc.

Reply:
> This drives me toward thinking that we ought not let users define livenessProbe and readinessProbe in their entirety, but should, instead, surface individual options for things like initialDelaySeconds, periodSeconds, timeoutSeconds, etc., etc.

I'd argue to go with my first implementation. With this, all options except the exec.command variables would be setable.

Reply:
> I'd argue to go with my first implementation. With this, all options except the exec.command variables would be setable.

Have a look at something like https://github.com/akuity/kargo/pull/3041 where we're looking to change how one of the probes is implemented.

Changes such as this would become much more difficult if we leave things _too_ configurable.